### PR TITLE
Allow multiple-config form for functions config in firebase.json

### PR DESCRIPTION
**Pull request by taeold**
This change extends `functions` config in `firebase.json` so that you can write it in multi-config form. E.g.

```json
// This has always been valid
{
  "functions": {
    "source": "source"
  }
}

// This is also valid now
{
  "functions": [
    {
      "source": "source"
    }
  ]
}
```

Since having multiple configuration for function doesn't make sense (yet), we add runtime checks to make sure that there is exactly 1 functions source if we have an array of function config. I had fun trying to add better types around configs and consolidated config validations to remove redundant config checks throughout the codebase.

### CODE REVIEW

File: src/deploy/functions/deploy.ts [hunk 1]
```diff
@@ -77,16 +77,10 @@ export async function deploy(
...
-      "Error: 'functions.source' is not defined"
-    );
+    const source = context.config.source;
     if (uploads.length) {
       logSuccess(
-        clc.green.bold("functions:") +
-          " " +
-          clc.bold(options.config.src.functions.source) +
-          " folder uploaded successfully"
+        clc.green.bold("functions:") + " " + clc.bold(source) + " folder uploaded successfully"
```
Comment (bkendall) [inline src/deploy/functions/deploy.ts:83]:
```suggestion
        `${clc.green.bold("functions:")} ${clc.bold(source)} folder uploaded successfully`
```

File: src/deploy/functions/prepare.ts [hunk 2]
```diff
@@ -127,12 +128,17 @@ export async function prepare(
...
+      runtimeConfig
+    );
   }
   if (backend.someEndpoint(wantBackend, (e) => e.platform === "gcfv2")) {
     context.functionsSourceV2 = await prepareFunctionsUpload(
-      /* runtimeConfig= */ undefined,
-      options
+      sourceDir,
+      context.config,
+      /* runtimeConfig= */ undefined
```
Comment (bkendall) [inline src/deploy/functions/prepare.ts:141]: Can be omitted if it's undefined.

File: src/deploy/functions/prepareFunctionsUpload.ts [hunk 3]
```diff
@@ -116,15 +114,9 @@ async function packageSource(options: Options, sourceDir: string, configValues:
 }
 
 export async function prepareFunctionsUpload(
-  runtimeConfig: backend.RuntimeConfigValues | undefined,
-  options: Options
+  sourceDir: string,
+  config: projectConfig.ValidatedSingle,
+  runtimeConfig: backend.RuntimeConfigValues | undefined
```
Comment (bkendall) [inline src/deploy/functions/prepareFunctionsUpload.ts:119]: Can use an optional argument

```suggestion
  runtimeConfig?: backend.RuntimeConfigValues
```

File: src/firebaseConfig.ts [hunk 4]
```diff
@@ -102,13 +102,14 @@ export type FirestoreConfig = {
...
 
-export type FunctionsConfig = {
-  // TODO: Add types for "backend"
+export type FunctionsSingle = {
   source?: string;
   ignore?: string[];
   runtime?: CloudFunctionRuntimes;
 } & Deployable;
 
+export type FunctionsConfig = FunctionsSingle | FunctionsSingle[];
```
Comment (bkendall) [inline src/firebaseConfig.ts:111]: I'm not sure if we have other guidance on this or not, but why not:

```typescript
export type FunctionConfig = {}

export type FunctionsConfig = FunctionConfig | FunctionConfig[];
```

Maybe it's a little "too close" with being plural or not, but "FunctionsSingle" seems... I dunno, I wouldn't know what to expect from that.

File: src/functions/projectConfig.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import { FirebaseError } from "../error";
+
+export type NormalizedConfig = [FunctionsSingle, ...FunctionsSingle[]];
+export type ValidatedSingle = FunctionsSingle & { source: string };
+export type ValidatedConfig = [ValidatedSingle];
+
+/**
+ * Normalize functions config to return functions config in an array form.
+ */
+export function normalize(config: FunctionsConfig | undefined): NormalizedConfig {
```
Comment (bkendall) [inline src/functions/projectConfig.ts:11]:
```suggestion
export function normalize(config?: FunctionsConfig): NormalizedConfig {
```

File: src/functions/projectConfig.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  if (config === undefined) {
```
Comment (bkendall) [inline src/functions/projectConfig.ts:12]: Does it need to specifically equal undefined?

```suggestion
  if (!config) {
```

File: src/functions/projectConfig.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  }
+  return [validateSingle(config[0])];
+}
+
+/**
+ * Normalize and validate functions config.
+ *
+ * Valid functions config has exactly one config and has all required fields set.
+ */
+export function normalizeAndValidate(config: FunctionsConfig | undefined): ValidatedConfig {
```
Comment (bkendall) [inline src/functions/projectConfig.ts:48]:
```suggestion
export function normalizeAndValidate(config?: FunctionsConfig): ValidatedConfig {
```

File: src/functions/projectConfig.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import { FunctionsConfig, FunctionsSingle } from "../firebaseConfig";
+import { FirebaseError } from "../error";
+
+export type NormalizedConfig = [FunctionsSingle, ...FunctionsSingle[]];
```
Comment (bkendall) [inline src/functions/projectConfig.ts:4]: So, it's `FunctionsSingle[]`? Why is a special config type needed?

File: src/serve/functions.ts [hunk 9]
```diff
@@ -9,6 +9,7 @@ import { parseRuntimeVersion } from "../emulator/functionsEmulatorUtils";
 import { needProjectId } from "../projectUtils";
 import { getProjectDefaultAccount } from "../auth";
 import { Options } from "../options";
+import * as functionsConfig from "../functions/projectConfig";
```
Comment (bkendall) [inline src/serve/functions.ts:12]: I find this rename confusing. Based on the file name, it should be `projectConfig`

```suggestion
import * as projectConfig from "../functions/projectConfig";
```

File: src/test/functions/projectConfig.spec.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
+import { expect } from "chai";
+
+import * as projectConfig from "../../functions/projectConfig";
+import { FirebaseError } from "../../error";
+
+const TEST_CONFIG = { source: "foo" };
+
+describe("projectConfig", () => {
+  describe("normalize", () => {
```
Comment (bkendall) [inline src/test/functions/projectConfig.spec.ts:9]: What if it's given more than one? Should that have a test?

**Review by bkendall — CHANGES_REQUESTED** [review]

File: src/firebaseConfig.ts (hunk 4 above)
Comment (taeold) [reply src/firebaseConfig.ts:111]: Lol. Naming things is hard. `Config` vs `Configs` was also a thing that was messing up my brain 🤔 .

Going ahead w/ your suggestion.

File: src/functions/projectConfig.ts (hunk 6 above)
Comment (taeold) [reply src/functions/projectConfig.ts:12]: This is probably fine!

File: src/functions/projectConfig.ts (hunk 8 above)
Comment (taeold) [reply src/functions/projectConfig.ts:4]: This type definition enforces that there be at least one item in the array. Too hard to read?

File: src/test/functions/projectConfig.spec.ts (hunk 10 above)
Comment (taeold) [reply src/test/functions/projectConfig.spec.ts:9]: That is checked in `validate`, and we have a test case for that.

File: src/functions/projectConfig.ts (hunk 8 above)
Comment (bkendall) [reply src/functions/projectConfig.ts]: It's maybe a little weird, but I guess I can live with it hah.

**Review by bkendall — APPROVED** [review]
Looks fine to me, but I think someone from functions should really take a look as well

File: src/functions/projectConfig.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+export type NormalizedConfig = [FunctionConfig, ...FunctionConfig[]];
+export type ValidatedSingle = FunctionConfig & { source: string };
+export type ValidatedConfig = [ValidatedSingle];
+
+/**
+ * Normalize functions config to return functions config in an array form.
+ */
+export function normalize(config?: FunctionsConfig): NormalizedConfig {
+  if (!config) {
+    throw new FirebaseError("No valid functions configuration detected in firebase.json");
```
Comment (colerogers) [inline src/functions/projectConfig.ts:13]: I think I'd prefer this error to be thrown before we call this function, like in `functions-config-export.ts` so we don't have an optional param. But this shouldn't block the PR

File: src/test/functions/projectConfig.spec.ts [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    it("normalizes singleton config", () => {
+      expect(projectConfig.normalize(TEST_CONFIG)).to.deep.equal([TEST_CONFIG]);
+    });
+
+    it("normalizes array config", () => {
+      expect(projectConfig.normalize([TEST_CONFIG])).to.deep.equal([TEST_CONFIG]);
+    });
+
+    it("throws error if given empty config", () => {
+      expect(() => projectConfig.normalize([])).to.throw(FirebaseError);
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:19]: nit - don't we usually check for the expected error message

File: src/test/functions/projectConfig.spec.ts [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+  });
+
+  describe("validate", () => {
+    it("passes validation for simple config", () => {
+      expect(projectConfig.validate([TEST_CONFIG])).to.deep.equal([TEST_CONFIG]);
+    });
+
+    it("fails validation given more than one config", () => {
+      expect(() =>
+        projectConfig.validate([TEST_CONFIG, { ...TEST_CONFIG, source: "bar" }])
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:30]: why don't you just write `projectConfig.validate([TEST_CONFIG, { source: "bar" }])`?

File: src/test/functions/projectConfig.spec.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      ).to.throw(FirebaseError);
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:31]: same with this error

File: src/test/functions/projectConfig.spec.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    it("fails validation given config w/o source", () => {
+      expect(() => projectConfig.validate([{ runtime: "nodejs10" }])).to.throw(FirebaseError);
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:35]: same with this error

File: src/test/functions/projectConfig.spec.ts [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    it("fails validation given config w/ empty source", () => {
+      expect(() => projectConfig.validate([{ source: "" }])).to.throw(FirebaseError);
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:39]: same with this error

File: src/test/functions/projectConfig.spec.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+      expect(projectConfig.normalizeAndValidate(TEST_CONFIG)).to.deep.equal([TEST_CONFIG]);
+    });
+
+    it("returns normalized config for multi-resource config", () => {
+      expect(projectConfig.normalizeAndValidate([TEST_CONFIG])).to.deep.equal([TEST_CONFIG]);
+    });
+
+    it("fails validation given singleton config w/o source", () => {
+      expect(() => projectConfig.normalizeAndValidate({ runtime: "nodejs10" })).to.throw(
+        FirebaseError
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:54]: same with this error

File: src/test/functions/projectConfig.spec.ts [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+      );
+    });
+
+    it("fails validation given singleton config w empty source", () => {
+      expect(() => projectConfig.normalizeAndValidate({ source: "" })).to.throw(FirebaseError);
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:59]: same with this error

File: src/test/functions/projectConfig.spec.ts [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    });
+
+    it("fails validation given multi-resource config w/o source", () => {
+      expect(() => projectConfig.normalizeAndValidate([{ runtime: "nodejs10" }])).to.throw(
+        FirebaseError
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:64]: same with this error

File: src/test/functions/projectConfig.spec.ts [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+      );
+    });
+
+    it("fails validation given more than one config", () => {
+      expect(() =>
+        projectConfig.normalizeAndValidate([TEST_CONFIG, { ...TEST_CONFIG, source: "bar" }])
+      ).to.throw(FirebaseError);
```
Comment (colerogers) [inline src/test/functions/projectConfig.spec.ts:71]: same with this error

File: src/config.ts [hunk 21]
```diff
@@ -76,13 +76,17 @@ export class Config {
...
+    if (this.projectDir && fsutils.dirExistsSync(this.path(Config.DEFAULT_FUNCTIONS_SOURCE))) {
+      if (Array.isArray(this.get("functions"))) {
+        if (!this.get("functions.[0].source")) {
+          this.set("functions.[0].source", Config.DEFAULT_FUNCTIONS_SOURCE);
+        }
+      } else {
+        if (!this.get("functions.source")) {
+          this.set("functions.source", Config.DEFAULT_FUNCTIONS_SOURCE);
+        }
+      }
```
Comment (colerogers) [inline src/config.ts:89]: Can we unwind this a bit?
```
if (!this.projectDir || !fsutils.dirExistsSync(this.path(Config.DEFAULT_FUNCTIONS_SOURCE))) {
    return;
}
if (Array.isArray(this.get("functions"))) {
    if (!this.get("functions.[0].source")) {
        this.set("functions.[0].source", Config.DEFAULT_FUNCTIONS_SOURCE);
    }
} else {
    if (!this.get("functions.source")) {
        this.set("functions.source", Config.DEFAULT_FUNCTIONS_SOURCE);
    }
}
```

Actually, I'm not sure if that makes it much easier to read

**Review by colerogers — APPROVED** [review]

File: src/functions/projectConfig.ts [hunk 22]
```diff
@@ -0,0 +1,50 @@
+import { FunctionsConfig, FunctionConfig } from "../firebaseConfig";
+import { FirebaseError } from "../error";
+
+export type NormalizedConfig = [FunctionConfig, ...FunctionConfig[]];
```
Comment (inlined) [inline src/functions/projectConfig.ts:4]: Is this the type definition of an array with at least one element?

File: src/firebaseConfig.ts [hunk 23]
```diff
@@ -102,13 +102,14 @@ export type FirestoreConfig = {
...
 
-export type FunctionsConfig = {
-  // TODO: Add types for "backend"
+export type FunctionConfig = {
   source?: string;
   ignore?: string[];
   runtime?: CloudFunctionRuntimes;
 } & Deployable;
 
+export type FunctionsConfig = FunctionConfig | FunctionConfig[];
```
Comment (inlined) [inline src/firebaseConfig.ts:111]: Function*s* vs Function config is incredibly subtle. I might prefer names as clear as FunctionsSingle | FunctionsMultiple

**Review by inlined — COMMENTED** [review]
I feel like I missed my baby's first steps. This is looking good and I'm excited to catch up.

**Pull request merged**